### PR TITLE
Add abstract hardware interface and refactor Robot Driver

### PR DESCRIPTION
**Pull request by jcnorby**
Changes include:
- Add HardwareInterface abstract class for extensibility
- Switch MBLinkConverter to SpiritInterface which inherits HardwareInterface
- Refactor robot_driver package to include directories for hardware interfaces and controllers
- Remove leg override node and support (deprecated)
- Remove open loop controller package (deprecated)
- Remove robot_proxy and ground_truth_publisher (deprecated)

These changes have been tested in sim but not yet in hardware. I would recommend we merge this PR sooner rather than later to permit additional refactoring, then test in hardware at a later date and submit PRs/hotfixes to resolve any remaining bugs.

### CODE REVIEW

File: robot_driver/src/controllers/leg_controller.cpp [hunk 1]
```diff
@@ -0,0 +1,38 @@
+#include "robot_driver/controllers/leg_controller.h"
```
Comment (jcnorby) [inline robot_driver/src/controllers/leg_controller.cpp:1]: Note this is not a new file, just renamed from leg_controller_template.cpp

File: robot_driver/include/robot_driver/hardware_interfaces/hardware_interface.h [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+   */
+  virtual bool send(const quad_msgs::LegCommandArray& leg_command_array_msg,
+                    const Eigen::VectorXd& user_data) = 0;
+
+  /**
+   * @brief Recieve data from the robot
+   * @param[out] joint_state_msg Message containing joint state information
+   * @param[out] imu_msg Message containing imu information
+   */
+  virtual bool recv(sensor_msgs::JointState& joint_state_msg,
```
Comment (justinyim) [inline robot_driver/include/robot_driver/hardware_interfaces/hardware_interface.h:62]: Should we just let users add additional variables if they want to use additional data from their robots (A1 foot contact sensors, other platforms' cameras or odometry handled by the built-in computers)?

File: robot_driver/src/robot_driver.cpp [hunk 3]
```diff
@@ -64,18 +65,19 @@ RobotDriver::RobotDriver(ros::NodeHandle nh, int argc, char** argv) {
...
   grf_pub_ = nh_.advertise<quad_msgs::GRFArray>(grf_topic, 1);
   leg_command_array_pub_ =
       nh_.advertise<quad_msgs::LegCommandArray>(leg_command_array_topic, 1);
   robot_heartbeat_pub_ =
       nh_.advertise<std_msgs::Header>(robot_heartbeat_topic, 1);
 
   // Set up pubs and subs dependent on robot layer
-  if (is_hw_) {
+  if (is_hardware_) {
     ROS_INFO("Loading hw robot driver");
```
Comment (justinyim) [inline robot_driver/src/robot_driver.cpp:81]: Expand hw to hardware?

File: robot_driver/include/robot_driver/hardware_interfaces/hardware_interface.h (hunk 2 above)
Comment (jcnorby) [reply robot_driver/include/robot_driver/hardware_interfaces/hardware_interface.h:64]: Added Eigen::VectorXd user_rx_data (in addition to user_tx_data).

**Review by justinyim — APPROVED** [review]
Looks good to me

**Comment by jcnorby** [comment]
Just tested and confirmed that this doesn't work in hardware yet, something causes mblink_.start(argc,argv) to segfault. Do we want to approve and fix it later or try to iron that out first?

**Comment by ardalantj** [comment]
Lets try and fix it if its segfaulting. I worry we wouldnt get to it later if its merged now.

**Comment by jcnorby** [comment]
These changes have now been tested in hardware, they require a clean catkin install so that catkin can handle the new paths correctly. I also updated the topic names for the imu and joint data, added a publisher for the joint data, and removed ground_truth_publisher and robot_proxy as they are no longer needed.

**Review by ardalantj — APPROVED** [review]

**Pull request merged**